### ISSUE TITLE
BatchNormalization with GPU placement (TensorFlow)

### ISSUE THREAD

**Issue by alrawi** [fchollet/keras#5802]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I have the following sample code:
https://gist.github.com/alrawi/0f6ce637d2955ab2e9c0d4b69509f5de

I cannot get BatchNormalization to be placed on GPU. My model fits on the GPU, but for BatchNormalization I get the following warning:

`WARNING:tensorflow:Tried to colocate gradients_2/batch_normalization_3/moments/sufficient_statistics/count_grad/Rank with an op batch_normalization_3/moments/sufficient_statistics/count that had a different device: /device:CPU:0 vs /device:GPU:2. Ignoring colocation property.`

I am suspecting that this is a type problem, since most GPUs don't support 64 bit operations. Here is some information about my system, I have 3 GPUs, device two is shown below:

`NVIDIA-SMI 367.48
Driver Version: 367.48
GeForce GTX TITAN`

Suggestions?

**Comment by vsenko** [comment]
I have a similar problem. Adding BatchNormalization causes the following output:
```
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Rank with an op batch_normalization_1/moments/sufficient_statistics/cou
nt that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/range_1/start with an op batch_normalization_1/moments/sufficient_stati
stics/count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/range_1/delta with an op batch_normalization_1/moments/sufficient_stati
stics/count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/range_1 with an op batch_normalization_1/moments/sufficient_statistics/
count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/ListDiff with an op batch_normalization_1/moments/sufficient_statistics
/count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/concat/axis with an op batch_normalization_1/moments/sufficient_statist
ics/count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/concat with an op batch_normalization_1/moments/sufficient_statistics/c
ount that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Gather with an op batch_normalization_1/moments/sufficient_statistics/c
ount that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Const with an op batch_normalization_1/moments/sufficient_statistics/co
unt that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Prod with an op batch_normalization_1/moments/sufficient_statistics/cou
nt that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Gather_1 with an op batch_normalization_1/moments/sufficient_statistics
/count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Const_1 with an op batch_normalization_1/moments/sufficient_statistics/
count that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
WARNING:tensorflow:Tried to colocate gradients/batch_normalization_1/moments/sufficient_statistics/count_grad/Prod_1 with an op batch_normalization_1/moments/sufficient_statistics/c
ount that had a different device: /device:CPU:0 vs /device:GPU:0. Ignoring colocation property.
```

**Comment by fchollet** [comment]
Can you provide us with the placement that TF does for every op in the BN
layer? These ops are created in a name scope by the name of the layer (in
Keras 2).

**Comment by alrawi** [comment]
Here is the placement log for the run above. I modified the code to do the placement log.

https://gist.github.com/alrawi/f92c1a6603134b2cc6b523b0c483d8b5

**Comment by Namnamseo** [comment]
This seems to be somehow realted to [an issue in TensorFlow](https://github.com/tensorflow/tensorflow/issues/3397). 

I'm not sure why this toy example works:

```
import tensorflow as tf
import numpy as np

with tf.device('/gpu:2'):
    X = tf.placeholder(tf.float32, shape=[None, 100])
    Y = tf.placeholder(tf.float32, shape=[None])
    mean, var = tf.nn.moments(X, axes=[0])
    norm = (X - mean) / var
    gamma = tf.Variable(tf.ones([1, 100]))
    beta = tf.Variable(tf.zeros([1, 100]))
    loss = tf.reduce_sum(tf.squared_difference(tf.reduce_sum(gamma*norm+beta, axis=[1]), Y))
    grad = tf.gradients(loss, [gamma, beta], colocate_gradients_with_ops=True)

config = tf.ConfigProto(log_device_placement=True, allow_soft_placement=False)
sess = tf.Session(config=config)

x = np.random.random((10, 100)).astype(np.float32)
y = np.random.random((10)).astype(np.float32)

with tf.device('/gpu:2'):
    sess.run(tf.global_variables_initializer())
    resres = sess.run(grad, feed_dict={X: x, Y: y})
print(resres)
```

**Comment by Namnamseo** [comment]
I was able to reproduce the same error with pure TensorFlow, so it seems the problem is in TensorFlow; `tf.moments()` and its gradient calculation on unresolved axis with GPU causes the error. I posted those results to the issue linked above.

**Comment by jonilaserson** [comment]
I have the same problem.

I created a model with a BatchNormalization layer.
I then do data parallelization - I split the batch into two halves, and
apply the model on each half (on different GPUs), then concatenate results.
Batchnormalization fails with the above error.

I'm assuming that it is because it doesn't know how to compute the moving
means across multiple GPUs?
How can I get around this problem?

There used to be a mode of BatchNormalization that does not compute a
moving mean but just takes the batch's mean.
Could it be utilized somehow?

**Comment by Namnamseo** [comment]
This is fixed in tensorflow/tensorflow#8906.
I tested @alrawi's sample code and monitored the device allocation. All of them are appropriately placed on GPU as intended. Device colocation error did not occur.

@fchollet, can we close it?

**Comment by fchollet** [comment]
Sure, SG
